man/man1/newspost.1
.TH "NEWSPOST" "1" "2.2.1" "Jim Faulkner" ""
.SH "NAME"
.LP 
newspost \- a usenet binary autoposter
.SH "SYNTAX"
.LP 
newspost [\fIoptions\fP [\fIarguments\fP]] <\fIfilenames\fP>
.SH "DESCRIPTION"
.LP 
Newspost is a usenet binary autoposter.  It will uuencode or yencode files
and post them to usenet newsgroups. In addition, it can automatically
generate and post .SFV checksum files and .PAR parity files.
.SH "OPTIONS"
.LP 
.TP 
\fB\-i\fR <\fIaddress\fP>
The news server to post to.  <\fIaddress\fP> must be either a hostname or
an IP address.  If the NNTPSERVER environment variable is set, this is set
to that value by default.
.TP 
\fB\-z\fR <\fInumber\fP>
Sets the port on which to connect to <\fInumber\fP>.
.TP 
\fB\-u\fR <\fIstring\fP>
Sets your username on the news server to <\fIstring\fP>.
.TP 
\fB\-p\fR <\fIstring\fP>
Sets your password on the news server to <\fIstring\fP>.
.TP 
\fB\-N\fR <\fIstring\fP>
Sets the amount of threads to use to <\fInumber\fP>.
.TP
\fB\-f\fR <\fIaddress\fP>
Your e\-mail address.  <\fIaddress\fP> must be a real e\-mail address, or
your posts may fail.  If the USER and HOSTNAME environment variables are
set, this is set to $USER@$HOSTNAME by default.
.TP 
\fB\-F\fR <\fIstring\fP>
Sets your name in the From header line to <\fIstring\fP>.  Please note:  
if your name has spaces in it, it should be enclosed in quotes.
.TP 
\fB\-o\fR <\fIstring\fP>
Sets the Organization header line to <\fIstring\fP>.  Please note:  if
your organization has spaces in it, it should be enclosed in quotes.
.TP 
\fB\-n\fR <\fIstring\fP>
Sets the newsgroup to post to.  <\fIstring\fP> may also be a comma
separated list of newsgroups, but you may only post to at most 5
newsgroups.
.TP 
\fB\-s\fR <\fIstring\fP>
Sets <\fIstring\fP> as the Subject header line.  This subject line is
always appended with ' \- filename (1/9)' (depending on the filename, part
number, and number of parts).  Other options may add or change information
in the subject line (see those options for details).  Please note:  if
your subject has spaces in it, it should be enclosed in quotes.  See
EXAMPLES for more information.
.TP 
\fB\-w\fR <\fIstring\fP>
Sets the Followup\-To header line.  <\fIstring\fP> must be a newsgroup.
.TP 
\fB\-m\fR <\fIaddress\fP>
Sets the Reply\-To header line.  <\fIaddress\fP> must be an e\-mail address.
.TP 
\fB\-r\fR <\fIstring\fP>
Set the References header line.  <\fIstring\fP> must be a message\-ID or a
comma\-separated list of message\-IDs.
.TP 
\fB\-x\fR
By default all posts include the "X\-No\-Archive: yes" header.  This
option DISABLES that behaviour.
.TP 
\fB\-X\fR <\fIstring\fP>
Sets <\fIstring\fP> as a complete header line.  This option may be
specified multiple times to add more header lines.  Please note:  do not
use this feature unless you know exactly what you are doing.
.TP 
\fB\-q\fR
This specifies that ' \- File 01 of 10: ' (depending on the file number and
number of files) will be added to the subject line after text specified by
the \-s switch, and before the file information.  If the file being posted
is a generated .SFV file, then the string is ' \- SFV File 1 of 1: '.  If
the file is a generated .PAR file, the string is ' \- PAR File 01 of 10: '
(depending on the .PAR file number and the number of .PAR files).  See
EXAMPLES for more information.
.TP 
\fB\-U\fR
Use uuencoding instead of yencoding.  Yencoding results in smaller, faster
posts, but archaic news readers do not support yencoding.
.TP 
\fB\-c\fR <\fIfilename\fP>
Generates and posts a .SFV checksum file named <\fIfilename\fP>.  If
<\fIfilename\fP> does not end in '.sfv', it will automatically be appended
to the filename.  Note the change from newspost 1.x, which used "\-v" for 
this option; also, the .SFV file is now posted first instead of last.
.TP 
\fB\-a\fR <\fIfilename\fP>
Generates and posts .PAR files whose name is based on <\fIfilename\fP>.  
If <\fIfilename\fP> does not end in '.par', it will automatically be 
appended to the filename.  The .PAR header will end in '.par', and 
each .PAR volume will end in '.p01', '.p02', etcetera.  Please see
http://parchive.sourceforge.net for more information about .PAR files.
.TP 
\fB\-A\fR <\fInumber\fP>
Sets the number of .PAR volumes to create to <\fInumber\fP>.  This is in
addition to the .PAR header file.  The \-a option must be used for this
option to do anything.
.TP 
\fB\-B\fR <\fInumber\fP>
Sets the number of .PAR volumes to create based on the number of files.  
For every <\fInumber\fP> files, newspost will create one .PAR volume.  
This is in addition to the .PAR header file.  By default, newspost creates
one .PAR volume for every ten files posted.
.TP 
\fB\-l\fR <\fInumber\fP>
Sets the number of lines per message to <\fInumber\fP>.  Most people post
messages which are between 3450 and 5000 lines long. By default, this is
set to 5000. Note: For uuencoded messages, this is the actual number of
lines in the body of the message; but for yencoded messages, it's used to 
determine the size of each segment before encoding, by multiplying the 
specified number of lines by 45 (which is the size of a uuencoded line 
before encoding). Thus, the size of each segment before encoding is the
same for either method, but the actual line count for yencoded segments 
will vary.
.TP 
\fB\-t\fR
When this option is specified, one file may be posted as a plain text
message.  If no file is specified, EDITOR (or vi) is opened to create the
message.
.TP 
\fB\-T\fR <\fInumber\fP>
Sets the time to wait before posting to <\fInumber\fP>.  This number is
interpreted in seconds by default, but you can specify, for example, "3m"
if you want 3 minutes, or "1h" if you want 1 hour.  This number can also
be specified in hours:minutes:seconds or minutes:seconds format.  e.g.
"5:30" if you want to wait 5 minutes 30 seconds, or "4:0:0" if you want to
wait 4 hours.  This value is set to 10 seconds by default, and may not be
set to less than 3 seconds.
.TP 
\fB\-k\fR <\fIdirname\fP>
Sets the directory to be used for storing temporary files to
<\fIdirname\fP>.  <\fIdirname\fP> must exist. If the TMPDIR environment
variable is set, this is set to that value by default.  If the TMP
environment variable is set (and TMPDIR is not), this is set to that value
by default.
.TP 
\fB\-d\fR
Sets the current options as default by saving them to $HOME/.newspostrc.  
Only the following options may be saved as default: \-i, \-z, \-u, \-p,
\-f, \-F, \-o, \-n, \-w, \-m, \-x, \-q, \-U, \-B, \-l, \-X, and \-k.
.TP 
\fB\-D\fR <\fIoption\fP>
This option disables or clears other options.  When U, q, or x is passed
as <\fIoption\fP>, it has the OPPOSITE effect of either the \-U, \-q, or
\-x option, respectively.  It will clear the corresponding options for u,
p, F, o, w, and m.  It also takes X as an argument, but it clears ALL
extra headers set with the \-X option.  This option may be specified
multiple times with different values for <\fIoption\fP>.  Please note only
options that are not required can be cleared.
.TP 
\fB\-v\fR
When this option is specified, newspost will be extra verbose.  (Note that 
"\-v" was used for .SFV generation in newspost 1.x.)
.TP 
\fB\-V\fR
Print version information and exit.
.TP 
\fB\-h\fR
Print help and exit.  This is the default when there are no arguments.
.SH "PARTIAL AND FAKE POSTING"
Parts of files may be posted by specifying filenames in the following
format: 'filename:partnumbers'.  Colons in the first or second position 
(as after a drive letter) are ignored.  Partnumbers is a comma\-separated list of
parts to post.  Each item in the list may be any of three formats: '2' (a
single part), '2\-5' (parts 2, 3, 4, and 5), or '2+' (part 2 and all parts
up to and including the last part).  If, for example, you want to post
parts 2, 4, 5, 6, 10, 11, and 12 of track01.mp3 (a 12 part file), you
could specify the filename like this: 'track01.mp3:2,4\-6,10+'.  If there
is no ':' in a specified filename, newspost will post the entire file. If
the number 0 is specified after the colon, newspost will add the file to generated .SFV
and .PAR files, and preserve ' \- File x of y: ' numbering, but not
actually post the file.  See EXAMPLES for more information.  Please note:  If your files
did not post completely, your file is much more likely to post properly if you repost 
the entire file, or split the file into smaller parts and post them.
.SH "SPEED INDICATOR"
While posting, Newspost shows the approximate upload speed in bytes per 
second.  This is measured by the total number of \fIencoded\fP bytes sent 
to the server.  In Newspost 1.x, the speed indicator was based on the number 
of \fIunencoded\fP bytes, so this speed indicator will show a higher number, 
though the speed is the same.  If you are connecting to your news server via 
a compressed link, you may see a difference in transfer rates when using 
uuencoding vs. yencoding.
.SH "FILES"
.LP 
\fI$HOME/.newspostrc\fP is an optional file used to store defaults. 
Newspost will also read (but not write) the old\-style .newspost file if it 
is present and .newspostrc is not.
.SH "ENVIRONMENT VARIABLES"
.LP 
.TP 
\fBHOME\fP
Specifies your home directory.  Newspost will not work properly if this is
not set.
.TP 
\fBNNTPSERVER\fP
Specifies the hostname or IP address of your news server.
.TP 
\fBUSER\fP
Used in creating the default from address if HOSTNAME is set.
.TP 
\fBHOSTNAME\fP
Used in creating the default from address if USER is set.
.TP 
\fBTMPDIR\fP
Specifies the directory used for creating temporary files.
.TP 
\fBTMP\fP
Specifies the directory used for creating temporary files if TMPDIR is not
set.
.TP 
\fBEDITOR\fP
Specifies the text editor to use.
.SH "EXAMPLES"
.LP 
.na
.TP 
\fBSave your news server, e\-mail address, and name as default:\fP
.br 
newspost \-d \-i news.myisp.com \-f newspost@sdf.lonestar.org \-F 'Jim Faulkner'
.TP 
\fBPost some files to alt.binaries.test:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs'
/nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some songs \- "track01.mp3" yEnc (01/12)
.TP 
\fBPost some files to alt.binaries.test using uuencoding:\fP
.br 
newspost \-U \-n alt.binaries.test \-s 'Here are some more songs' /nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some songs \- track01.mp3 (01/15)
.TP 
\fBInclude "File x of y" in the subject:\fP
.br 
newspost \-q \-n alt.binaries.test \-s 'Here are some more songs'
track01.mp3 track02.mp3 track03.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some more songs \- File 1 of 3: "track01.mp3" yEnc (01/12)
.TP 
\fBMake uuencode the default:\fP
.br 
newspost \-d \-U
.TP 
\fBAdd extra headers to every post by default:\fP
.br 
newspost \-d \-X 'X\-My\-Stupid\-Header\-1' \-X 'X\-My\-Stupid\-Header\-2'
.TP 
\fBClear extra headers and "File x of y", set uuencode, default:\fP
.br 
newspost \-d \-DX \-Dq \-U
.TP 
\fBGenerate and include an SFV file:\fP
.br 
newspost \-c checksum.sfv \-n alt.binaries.test \-s 'Here are some songs'
/nfs/music/*.mp3
.TP 
\fBGenerate and include PAR files:\fP
.br 
newspost \-a parfile.par \-n alt.binaries.test \-s 'Here are some songs'
/nfs/music/*.mp3
.TP 
\fBCreate a PAR volume for every 7 files posted by default:\fP
.br 
newspost \-d \-B 7
.TP 
\fBPost with 3 PAR volumes:\fP
.br 
newspost \-q \-a parfile.par \-A 3 \-n alt.binaries.test \-s 'Here are
some songs' /nfs/music/*.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
Here are some songs \- PAR File 1 of 4: "parfile.par" yEnc (1/1)
.TP 
\fBPost parts 3, 4, and 5 of track01.mp3 to alt.binaries.test:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs' track01.mp3:3\-5
.TP 
\fBPost all of track01.mp3, parts of track02.mp3 and track03.mp3:\fP
.br 
newspost \-n alt.binaries.test \-s 'Here are some songs' track01.mp3
track02.mp3:11+ track03.mp3:1,3,6\-8
.TP 
\fBResume a broken post at Track02.mp3 preserving -q numbering:\fP
.br 
newspost \-q \-s '2 songs' Track01.mp3:0 Track02.mp3
.TP 
\fBA subject line from the above post may look like this:\fP
.br 
2 songs \- File 2 of 2: "Track02.mp3" yEnc (1/8)
.TP 
\fBPost an .SFV file and Track02.mp3, but include Track01.mp3 in the .SFV file:\fP
.br 
newspost \-c songs.SFV \-s '2 songs' Track01.mp3:0 Track01.mp3
.ad
.SH "AUTHORS"
.LP 
.ad
Mostly written by Jim Faulkner <newspost@sdf.lonestar.org>, with some help from 
William McBrine <wmcbrine@users.sf.net>. Portions borrowed from cksfv by 
Bryan Call, and parchive by Willem Monsuwe.
Threading and some small changes by Pietje Bell <pietjebell@ana-chan.com>
.SH "AVAILABILITY"
The most recent version of newspost is available at
http://github.com/PietjeBell88/newspost


CHANGES
Newspost 

List of Changes

2.2.1
 - Added the ability to use threads for binary posting. Use -N <int> to
specifiy the number of threads to use.

2.1.1
 - The extra header option (-X) works again
 - Subject line creation is fixed when ALLOW_NO_SUBJECT is defined
 - The text prefix is no longer counted in the "UUencoding xx bytes" line
 - Updated the contact e-mail address

2.1
 - Removed all string length limitations
 - Fixed zero-padding with -q option ("File 01 of 10" not "File 1 of 10")
 - Referencing message IDs (-r option) now works
 - Ability to "fake post" files: newspost can add specific files to 
generated .SFV and .PAR files, and preserve ' \- File x of y: ' numbering, 
but not actually post those files
 - Compilation fixes for a variety of platforms

2.0 - Major rewrite
 - Yencoding support
 - Support for automatic PAR file generation
 - Support for posting parts of files
 - Optionally edit text prefixes and text posts in external editor
 - User-configurable delay before posting
 - New header options, including user-specified extra header lines
 - New file format for defaults, now in .newspostrc
 - Man page :-)
 - SFV posted first instead of last
 - SFV now specified with "-c" instead of "-v"; now "-v" is "verbose"

1.20 - changes by William McBrine
 - .newspost values were sometimes not zero-terminated on rereading,
resulting in garbled settings; fixed
 - messages that had a '.' in the first position on a line could be
posted incorrectly; fixed
 - changed status display to show "Posting n of n" instead of hash
marks
 - lines in the body of the article now end correctly in CRLF,
instead of just LF
 - temporary files are no longer created, except in the case of SFVs
generated by newspost
 - CRLF vs. LF is now handled correctly in reading and writing files
(and should now work under Windows as well as Unix)
 - "Organization" line is now optional; "Subject" parameter is 
optionally optional (see source)
 - corrected format of "User-Agent" line
 - special makefiles for AIX and OSF no longer needed
 - much internal reorganization; faster, smaller, and probably easier
to compile

1.13
 - fixed endless loop that occurs with -v option when one of the
files is deleted before posting is complete
 - no longer prints out your password when displaying help
 - added a makefile and made changes to support DEC alpha OSF1 V4.0 
(thanks Chris Gentle)
 - added a makefile for AIX
 - documentation updated

1.12
 - file permissions on ~/.newspost are now changed to 600 when
it is created or modified
 - fixed bug in creating headers when -t option is used
 - homepage is now included in the user-agent line
 - new -V switch to check the version

1.11
 - the number of digits in the part number field matches the number
of digits in the total part count (thanks Mike Liang)

1.1
 - no longer uses uulib
 - added a makefile for solaris
 - now can reply to messages using -r
 - misc bugfixes

1.01a
 - fixed typo in newspost -h

1.01
 - fixed problems with some news servers

1.0
 - initial release



base/newspost.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/*
 * Authors: Jim Faulkner <newspost@sdf.lonestar.org>
 *          and William McBrine <wmcbrine@users.sf.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 */

#include <pthread.h>
#include <sys/time.h>
#include "newspost.h"
#include "queue.h"
#include "../ui/ui.h"
#include "socket.h"
#include "nntp.h"
#include "encode.h"
#include "../cksfv/sfv.h"
#include "../parchive/parintrf.h"

/**
*** Private Declarations
**/

typedef struct {
	newspost_data *data;
	newspost_threadinfo *threadinfo;
	queue *fifo;
}
poster_thread_arg;

static int post_text_file(newspost_data *data, SList *file_list);

static int encode_and_post(newspost_data *data, SList *file_list,
			    SList *parfiles);

static void post_file(newspost_data *data, queue *fifo, file_entry *file_data,
		      int filenumber, int number_of_files, const char *filestring);

static void *poster_thread(void *arg);

static Buff *make_subject(Buff *subject, newspost_data *data,
	int filenumber, int number_of_files, const char *filename,
	int partnumber, int number_of_parts, const char *filestring);

static SList *preprocess(newspost_data *data, SList *file_list);

static Buff *read_text_file(Buff * text_buffer, const char *filename);

/**
*** Public Routines
**/

int newspost(newspost_data *data, SList *file_list) {
	int retval;
	SList *parfiles = NULL;

	/* preprocess */
	if (data->text == FALSE)
		parfiles = preprocess(data, file_list);

	/* and post! */
	ui_post_start(data, file_list, parfiles);

	if (data->text == TRUE)
		retval = post_text_file(data, file_list);
	else
		retval = encode_and_post(data, file_list, parfiles);

	return retval;
}


/**
*** Private Routines
**/

/* Text Posting */
static int post_text_file(newspost_data *data, SList *file_list) {
	file_entry *file_data = NULL;
	int retval = NORMAL;
	Buff *text_buffer = NULL;
	newspost_threadinfo tinfo;

	/* set the threadinfo */
	tinfo.bytes_written = 0;
	tinfo.rwlock = (pthread_rwlock_t *) malloc(sizeof(pthread_rwlock_t));
	pthread_rwlock_init(tinfo.rwlock, NULL);
	tinfo.thread_id = 1;

	/* create the socket */
	ui_socket_connect_start(&tinfo, data->address->data);
	tinfo.sockfd = socket_create(data->address->data, data->port);
	retval = tinfo.sockfd;
	if (retval < 0)
		return retval;

	ui_socket_connect_done(&tinfo);

	/* log on to the server */
	ui_nntp_logon_start(&tinfo, data->address->data);
	if (nntp_logon(&tinfo, data) == FALSE) {
		socket_close(tinfo.sockfd);
		return LOGON_FAILED;
	}
	ui_nntp_logon_done(&tinfo);

	file_data = file_list->data;
	/* post */
	text_buffer = read_text_file(text_buffer, file_data->filename->data);
	if(text_buffer != NULL)
		retval = nntp_post(&tinfo, data->subject->data, data, text_buffer->data,
					text_buffer->length, TRUE);

	pthread_rwlock_destroy(tinfo.rwlock);
	free(tinfo.rwlock);
	buff_free(text_buffer);

	return retval;
}

/* Binary Posting (thread-based): */
static int encode_and_post(newspost_data *data, SList *file_list,
			    SList *parfiles) {
	int number_of_files;
	int i, j;
	file_entry *file_data = NULL;
	int retval = NORMAL;
	pthread_t *thread_array =
		(pthread_t *) malloc(data->threads * sizeof(pthread_t));
	newspost_threadinfo *threadinfo_array =
		(newspost_threadinfo *) malloc(data->threads * sizeof(newspost_threadinfo));
	poster_thread_arg *poster_args =
		(poster_thread_arg *) malloc(data->threads * sizeof(poster_thread_arg));
	queue *fifo;

	fifo = queue_init(data->threads * 2);

	for(j = 0; j < data->threads; j++) {
		/* set the thread's local data */
		threadinfo_array[j].thread_id = j+1;
		threadinfo_array[j].status = THREAD_INITIALIZING;
		threadinfo_array[j].bytes_written = 0;
		threadinfo_array[j].rwlock = (pthread_rwlock_t *) malloc(sizeof(pthread_rwlock_t));
		pthread_rwlock_init(threadinfo_array[j].rwlock, NULL);

		poster_args[j].data = data;
		poster_args[j].threadinfo = &threadinfo_array[j];
		poster_args[j].fifo = fifo;

		/* create a poster thread */
		pthread_create(&thread_array[j], NULL, poster_thread, &poster_args[j]);
	}

	/* post any sfv files... */
	if (data->sfv != NULL) {
		file_data = file_entry_alloc(file_data);
		file_data->filename =
			buff_create(file_data->filename, "%s", data->sfv->data);
		if (stat(data->sfv->data, &file_data->fileinfo) == -1)
			ui_sfv_gen_error(data->sfv->data, errno);
		else {
			post_file(data, fifo, file_data, 1, 1, "SFV File");

			unlink(data->sfv->data);
		}
		free(file_data);
	}

	number_of_files = slist_length(file_list);

	/* post the files */
	i = 1;
	while (file_list != NULL) {

		file_data = (file_entry *) file_list->data;

		post_file(data, fifo, file_data, i, number_of_files, "File");

		i++;
		file_list = slist_next(file_list);
	}

	/* post any par files */
	i = 1;
	file_list = parfiles;
	number_of_files = slist_length(parfiles);
	while (file_list != NULL) {

		file_data = (file_entry *) file_list->data;

		post_file(data, fifo, file_data, i, number_of_files, "PAR File");

		unlink(file_data->filename->data);
		buff_free(file_data->filename);
		free(file_data);
		i++;
		file_list = slist_next(file_list);
	}
	slist_free(parfiles);

	/* Signal that there will no new items be written to the queue */
	pthread_mutex_lock(fifo->mut);
	fifo->producer_done = TRUE;
	pthread_mutex_unlock(fifo->mut);

	/* Signal that the producer is done */
	pthread_cond_broadcast(fifo->cond_producer_done);

	/* Wait for the poster threads to clear the queue */
	pthread_mutex_lock(fifo->mut);
	pthread_cond_wait(fifo->cond_empty, fifo->mut);
	pthread_mutex_unlock(fifo->mut);

	/* wake-up sleeping threads so they can exit */
	pthread_cond_broadcast(fifo->cond_not_empty);

	for(j = 0; j < data->threads; j++) {
		pthread_rwlock_rdlock(threadinfo_array[j].rwlock);
		if (threadinfo_array[j].status == THREAD_CONNECTING)
			pthread_cancel(thread_array[j]);
		pthread_rwlock_unlock(threadinfo_array[j].rwlock);

		pthread_join(thread_array[j], NULL);
	}

	queue_delete(fifo);
	for(j = 0; j < data->threads; j++) {
		pthread_rwlock_destroy(threadinfo_array[j].rwlock);
		free(threadinfo_array[j].rwlock);
	}
	free(threadinfo_array);
	free(poster_args);
	free(thread_array);

	return retval;
}

static void post_file(newspost_data *data, queue *fifo, file_entry *file_data,
		      int filenumber, int number_of_files, const char *filestring) {
	int j;
	int number_of_parts =
		get_number_of_encoded_parts(data, file_data);
	Buff * subject = NULL;
	post_article_t article;

	if(file_data->parts != NULL){
		if(file_data->parts[0] == TRUE) return;
	}

	for (j = 1; j <= number_of_parts; j++) {

		if ((file_data->parts != NULL) &&
		    (file_data->parts[j] == FALSE))
			continue;

		subject = make_subject(subject, data, filenumber, number_of_files,
			     file_data->filename->data, j, number_of_parts,
			     filestring);
		
		article.file_data = file_data;
		article.partnumber = j;
		article.subject = subject;

		/* Add item to queue */
		pthread_mutex_lock(fifo->mut);
		while (fifo->full)
			pthread_cond_wait(fifo->cond_not_full, fifo->mut);

		queue_item_add(fifo, &article);

		pthread_mutex_unlock(fifo->mut);
		pthread_cond_signal(fifo->cond_not_empty);
	}
	buff_free(subject);

	return;
}

static void *poster_thread(void *arg)
{
	/* readability */
	poster_thread_arg *arguments = (poster_thread_arg *) arg;

	newspost_data *data = arguments->data;
	newspost_threadinfo *tinfo = arguments->threadinfo;
	queue *fifo = arguments->fifo;

	/* variable declaration/definition */
	post_article_t article;
	char *data_buffer;

	int total_failures = 0;
	int number_of_tries = 0;

	int retval;

	int number_of_bytes;

	struct timespec   ts;
	struct timeval    tp;

	tinfo->sockfd = -1;

	/* initialize */
	article.file_data = NULL;
	article.partnumber = -1;
	article.subject = NULL;

	pthread_rwlock_wrlock(tinfo->rwlock);
	tinfo->status = THREAD_CONNECTING;
	pthread_rwlock_unlock(tinfo->rwlock);

	/* create the socket */
	while (tinfo->sockfd < 0) {
		ui_socket_connect_start(tinfo, data->address->data);
		tinfo->sockfd = socket_create(data->address->data, data->port);

		if (tinfo->sockfd >= 0)
			break;

		ui_socket_connect_failed(tinfo, tinfo->sockfd);

		gettimeofday(&tp, NULL);

		/* Convert from timeval to timespec */
		ts.tv_sec  = tp.tv_sec;
		ts.tv_nsec = tp.tv_usec * 1000;
		ts.tv_sec += SOCKET_RECONNECT_WAIT_SECONDS;

		number_of_tries++;

		if (number_of_tries >= 5) {
			ui_connecting_too_many_failures(tinfo);
			pthread_exit(NULL);
		}
		pthread_mutex_lock(fifo->mut);

		while(!fifo->producer_done) {
			retval = pthread_cond_timedwait(fifo->cond_producer_done, fifo->mut, &ts);
			if (retval == ETIMEDOUT || retval == 0)
				break;
		}

		/* quit if the producer signalled it was done */
		if (fifo->producer_done) {
			pthread_mutex_unlock(fifo->mut);
			pthread_exit(NULL);
		}
		pthread_mutex_unlock(fifo->mut);
	}
	ui_socket_connect_done(tinfo);

	pthread_rwlock_wrlock(tinfo->rwlock);
	tinfo->status = THREAD_POSTING;
	pthread_rwlock_unlock(tinfo->rwlock);

	number_of_tries = 0;

	/* log on to the server */
	ui_nntp_logon_start(tinfo, data->address->data);
	if (nntp_logon(tinfo, data) == FALSE) {
		socket_close(tinfo->sockfd);
		pthread_exit(NULL);
	}
	ui_nntp_logon_done(tinfo);

	/* allocate the buffer */
	data_buffer = (char *) malloc(get_buffer_size_per_encoded_part(data));

	while (TRUE) {
		pthread_mutex_lock(fifo->mut);
		while (fifo->empty && !fifo->producer_done)
			pthread_cond_wait(fifo->cond_not_empty, fifo->mut);

		retval = queue_item_del(fifo, &article);

		pthread_mutex_unlock(fifo->mut);

		if (retval == QUEUE_PRODUCER_DONE) {
			pthread_cond_broadcast(fifo->cond_empty);
			break;
		}

		pthread_cond_signal(fifo->cond_not_full);

		number_of_bytes = get_encoded_part(data, article.file_data, article.partnumber, data_buffer);

		if (article.partnumber == 1)
			ui_posting_file_start(data, article.file_data);

		ui_posting_part_start(tinfo, article.file_data, article.partnumber);

		retval = nntp_post(tinfo, article.subject->data, data, data_buffer, number_of_bytes, FALSE);

		if (retval == NORMAL) {
			ui_posting_part_done(tinfo, article.file_data, article.partnumber);

			/* check if this part was the last one of a certain file */
			pthread_rwlock_wrlock(article.file_data->rwlock);
			if (++article.file_data->parts_posted == article.file_data->parts_to_post)
				ui_posting_file_done(data, article.file_data);
			pthread_rwlock_unlock(article.file_data->rwlock);
		}
		else if (retval == POSTING_NOT_ALLOWED)
			return NULL;
		else {
			if (number_of_tries < 5) {
				ui_nntp_posting_retry(tinfo);
				sleep(5);
				number_of_tries++;
				continue;
			}
			else {
				total_failures++;
				if (total_failures == 5) {
					nntp_logoff(tinfo);
					socket_close(tinfo->sockfd);
					ui_posting_too_many_failures(tinfo);
				}
			}
		}
		number_of_tries = 0;
	}

	nntp_logoff(tinfo);
	socket_close(tinfo->sockfd);

	buff_free(article.subject);
	free(data_buffer);

	pthread_rwlock_wrlock(tinfo->rwlock);
	tinfo->status = THREAD_DONE;
	pthread_rwlock_unlock(tinfo->rwlock);

	pthread_exit(NULL);
	return NULL;
}

static Buff *make_subject(Buff *subject, newspost_data *data, int filenumber,
			 int number_of_files, const char *filename,
			 int partnumber, int number_of_parts,
			 const char *filestring) {
	char numbuf[32];
	int numsize;

	if (data->subject != NULL)
		subject = buff_create(subject, "%s - ", data->subject->data);
	else
		subject = buff_create(subject, "");
	if (data->filenumber == TRUE){
		sprintf(numbuf, "%i", number_of_files);
		numsize = strlen(numbuf);
		subject = buff_add(subject, "%s %0*i of %i: ", filestring,
				   numsize, filenumber, number_of_files);
	}
	sprintf(numbuf, "%i", number_of_parts);
	numsize = strlen(numbuf);
	subject = buff_add(subject, (data->uuenc == FALSE) ? "\"%s\" yEnc (%0*i/%i)" :
		"%s (%0*i/%i)", n_basename(filename), numsize,
		partnumber, number_of_parts);
	return subject;
}

static SList *preprocess(newspost_data *data, SList *file_list) {
	Buff *tmpstring = NULL;
	SList *parfiles = NULL;

	/* make the from line */
	if (data->name != NULL) {
		tmpstring = buff_create(tmpstring, "%s", data->from->data);
		data->from = buff_create(data->from, "%s <%s>",
					 data->name->data, tmpstring->data);
		buff_free(tmpstring);
	}

	/* calculate CRCs if needed; generate any sfv files */
	if ((data->uuenc == FALSE) || (data->sfv != NULL)) {
		calculate_crcs(file_list);

		if (data->sfv != NULL)
			newsfv(file_list, data);
	}

	/* generate any par files */
	if (data->par != NULL) {
		parfiles = par_newspost_interface(data, file_list);
		if (data->uuenc == FALSE)
			calculate_crcs(parfiles);
	}

	return parfiles;
}

/* returns number of bytes read */
static Buff *read_text_file(Buff *text_buffer, const char *filename) {
	FILE *file;
	Buff *line = NULL;

	buff_free(text_buffer);
	file = fopen(filename, "r");
	if (file != NULL) {
		while (!feof(file)) {
			line = buff_getline(line, file);
			if(line == NULL){
				text_buffer = buff_add(text_buffer, "\r\n");
				continue;
			}

			/* translate for posting */
			if (line->data[0] == '.')
				text_buffer = buff_add(text_buffer, ".");

			text_buffer = buff_add(text_buffer, "%s", line->data);

			if(text_buffer->data[(text_buffer->length - 1)] == '\r')
				text_buffer = buff_add(text_buffer, "\n");
			else
				text_buffer = buff_add(text_buffer, "\r\n");
		}
		fclose(file);
	}
	return text_buffer;
}
